Tradução: data-transform.qmd (#97)
cienciadedatos/pt-r4ds: data-transform.qmd

  )
```

Verbos do dplyr são organizados em quatro grupos baseados na estrutura sobre a qual operam: **linhas** (_rows_), **colunas** (_columns_), **grupos** (_groups_) ou **tabelas** (_tables_).
Nas próximas seções você aparenderá os verbos mais importantes para linhas, colunas e grupos, então voltaremos para os verbos de _join_ que operam em tabelas na seção @sec-joins.
Vamos nessa!

## Linhas (rows)

Os verbos mais importantes que operam nas linhas de um conjunto de dados são `filter()`, que modifica quais linhas estão presentes sem mudar a sua ordem, e `arrange()`, que modifica a ordem das linhas sem modificar quais estão presentes.

> Ambas as funções afetam apenas as linhas, e as colunas permanecem intocadas.

Reviewer: ```suggestion
Ambas as funções afetam apenas as linhas e as colunas permanecem intocadas.
```

Também vamos falar sobre `distinct()`, que acha as linhas com valores únicos, mas diferentemente de `arrange()` e `filter()` também pode, opcionalmente, modificar as colunas.

### `filter()`

`filter()` lhe permite manter linhas com base nos valores das colunas[^data-transform-1].
O primeiro argumento é o data frame.
O segundo e demais argumentos são as condições que devem ser verdadeiras para manter a linha.
Por exemplo, podemos achar todos os vôos que decolaram com mais de 120 minutos (duas horas) de atraso:

[^data-transform-1]: Depos você irá aprender sobre a família de funções `slice_*()` que permite escolher linhas com base em suas posições.

```{r}
flights |> 
  filter(dep_delay > 120)
```

> Assim como `>` (maior que), você também pode usar `>=` (maior ou igual que), `<` (menor que), `<=` (menor ou igual que), `==` (igual), e `!=` (diferente que).

Reviewer: ```suggestion
Assim como `>` (maior que), você também pode usar `>=` (maior ou igual a), `<` (menor que), `<=` (menor ou igual a), `==` (igual a), e `!=` (diferente de).
```

  filter(month == 1 | month == 2)
```

Existe um atalho útil quando você está combinando `|` e `==`: `%in%`.
Ele vai manter as linhas cujas variáveis são iguais a um dos valores à direita:

```{r}
# Uma forma mais curta de selecionar vôos que decolaram em janeiro ou fevereiro
flights |> 
  filter(month %in% c(1, 2))
```

Vamos voltar para essas comparações e operadores lógicos com mais detalhes em @sec-logicals.

Quando você roda `filter()` o dplyr executa a operação de filtragem, criando um novo data frame, e em seguida o exibe. 

> Ele não modifica o conjunto de dados `flights` já existente, pois as funções do dplyr nunca modificam seus argumentos de entrada.

Author: Oi @Luaaalex !

Creio que aqui usar apenas "entrada" pode ficar um pouco confuso em português. Acho que na versão originao, `inputs` cria um contexto que deixa pouca margem para ambiguidade, mas em português, só "entrada" fica meio vago, por isso optei por adicionar "argumentos" nesse caso para deixar mais claro que estamos falando sobre os argumentos da função.

O que acha?

> Ele não modifica o conjunto de dados `flights` já existente, pois as funções do dplyr nunca modificam seus argumentos de entrada.

Reviewer: Concordo, @halian-vilela

  filter(month == 1 | 2)
```

> Isso até "funciona", no sentido de que não dará um erro, mas ele não vai fazer o que você esperaria, pois `|` primeiro vai avaliar a condição `month == 1` e depois a condição `2`, que não é uma condição adequada para se avaliar nesse caso.

Reviewer: ```suggestion
Isso até "funciona", no sentido de que não retornará um erro, mas ele não vai fazer o que você esperaria, pois `|` primeiro vai avaliar a condição `month == 1` e depois a condição `2`, que não é uma condição adequada para se avaliar nesse caso.
```

```{r}
flights |> 
  distinct(origin, dest, .keep_all = TRUE)
```

Não é uma coincidência que todos esses vôos distintos sejam em 1º de janeiro: `distinct()` irá achar a primeira ocorrência de uma linha única no conjunto de dados e decartar todas as demais.

Se, em vez disso, você quiser achar o número de ocorrências, é melhor trocar `distinct()` por `count()`, e com o argumento `sort = TRUE` você pode organizá-los em ordem decrescente do número de ocorrências. 
Você vai aprendar mais sobre contagens em @sec-counts.

```{r}
flights |>
  count(origin, dest, sort = TRUE)
```

### Exercícios

1.  Utilizando um único pipeline para cada condição, ache todos os vôos que cumpram a condição: 

    -   Teve um atraso na chegada de duas ou mais horas 
    -   Voou para Houston (`IAH` ou `HOU`)
    -   Foi operado pela United, American ou Delta 
    -   Decolou no verão^[no hemisfério norte] (julho, agosto e setembro)
    -   Chegou com mais de duas horas de atraso, mas não teve atraso na decolagem. 
    -   Atrasou pelo menos uma hora, mas recuperou mais de 30 minutos em vôo. 

2.  Ordene `flights` para achar os vôos com os maiores atrasos na decolagem. 

>     Ache os vôos que saíram o mais cedo na manhã. 

Reviewer: ```suggestion
    Ache os vôos que saíram o mais cedo pela manhã. 
```

3.  Ordene `flights` para achar os vôos mais rápidos.
    (Dica: tente incluir algum cálculo matemático dentro da sua função.)

> 4.  Teve pelo menos um vôo em cada dia de 2013?

Reviewer: ```suggestion
4.  Houve pelo menos um vôo em cada dia de 2013?
```

  )
```

Ou então, você pode controlar quais variáveis são mantidas com o argumento `.keep`. 
Um argumento particularmente útil é `"used"` que especifica que nós queremos manter apenas as colunas que estiveram envolvidas ou foram criadas pela etapa de `mutate()`.

> Por exemplo, a saída a seguir irá conter apenas as variáveis `dep_delay`, `arr_delay`, `air_time`, `gain`, `hours` e `gain_per_hour`.

Reviewer: ```suggestion
Por exemplo, a saída a seguir conterá apenas as variáveis `dep_delay`, `arr_delay`, `air_time`, `gain`, `hours` e `gain_per_hour`.
```

batters
```

> Quando plotamos a habilidade dos rebatedores (medidas pela média de acerto, `performance`) contra o número de oportunidades de acertar a bola (medidas pela númer de vezes ao bastão, `n`), é possível ver dois padrões:

Reviewer: Aqui eu tenho quase certeza de que a @beatrizmilz tinha uma sugestão para não usar esse verbo emprestado. Procurei, mas não achei no guia. Considere algo como

"Quando montamos um gráfico da habilidade dos rebatedores (...) contra o número de oportunidades (...)"  ou número de oportunidades em função da habilidade dos rebatedores.

Author: ```suggestion
Quando montamos um gráfico da habilidade dos rebatedores (medidas pela média de acerto, `performance`) contra o número de oportunidades de acertar a bola (medidas pela númer de vezes ao bastão, `n`), é possível ver dois padrões:
```

1.  A variação em `performance` é maior entre os jogadores que foram menos vezes ao bastão. 

>     A forma do gráfico é muito característica: sempre que você plotar uma média (ou alguma outra sumarização estatística)  vs. tamanho do grupo, você verá que a variação diminui à medida que o tamanho da amostra aumenta[^data-transform-4].

Reviewer: ```suggestion
    A forma do gráfico é muito característica: sempre que você projetar no gráfico uma média (ou alguma outra sumarização estatística)  vs. tamanho do grupo, você verá que a variação diminui à medida que o tamanho da amostra aumenta[^data-transform-4].
```

2.  Existe uma correlação positiva entre a habilidade (`performance`) e as oportunidades de acertar a bola (`n`), pois os times querem dar aos melhores jogadores o maior número de oportunidades de acertar a bola.

[^data-transform-4]: \*cof-cof\* a lei dos grandes números \*cof-cof\*.

```{r}
#| warning: false
#| fig-alt: |

> #|   Um gráfico de dispersão do número de performance de rebatidas vs. oportunidades de rebatidas 

Reviewer: ```suggestion
#|   Um gráfico de dispersão do performance de rebatidas vs. oportunidades de rebatidas 
```

#|   sobrepostas por uma linha suavizada. Performance média cresce agudamente 
#|   a partir de 0.2 quando n é 1 até 0.25 quando n é ~1000. Performance média 

> #|   continua crescendo linearmente em uma inclinação muito mais rasa atingindo 

Reviewer: ```suggestion
#|   continua crescendo linearmente em uma inclinação muito mais suave atingindo 
```

batters
```

Quando plotamos a habilidade dos rebatedores (medidas pela média de acerto, `performance`) contra o número de oportunidades de acertar a bola (medidas pela númer de vezes ao bastão, `n`), é possível ver dois padrões:

1.  A variação em `performance` é maior entre os jogadores que foram menos vezes ao bastão. 

>     A forma do gráfico é muito característica: sempre que você plotar uma média (ou alguma outra síntese estatística)  vs. tamanho do grupo, você verá que a variação diminui à medida que o tamanho da amostra aumenta[^data-transform-4].

Reviewer 2: ```suggestion
    A forma do gráfico é muito característica: sempre que você plotar uma média (ou alguma outra sumarização estatística)  vs. tamanho do grupo, você verá que a variação diminui à medida que o tamanho da amostra aumenta[^data-transform-4].
```
